lib/std/fs/test.zig

```diff
@@ -1404,3 +1404,69 @@ test "File.PermissionsUnix" {
     try testing.expect(permissions_unix.unixHas(.user, .execute));
     try testing.expect(!permissions_unix.unixHas(.other, .execute));
 }
+
+test "preopens" {
+    if (builtin.os.tag != .wasi) return error.SkipZigTest;
+
+    // lifted from `testing`
+    const random_bytes_count = comptime (12);
+    const sub_path_len = comptime (std.fs.base64_encoder.calcSize(random_bytes_count));
+    const path = comptime ("/tmp");
+    const tmp_file_name = comptime ("file.txt");
+    const nonsense = comptime ("nonsense");
+
+    var random_bytes: [random_bytes_count]u8 = undefined;
+    var sub_path: [sub_path_len]u8 = undefined;
+
+    std.crypto.random.bytes(&random_bytes);
+    _ = std.fs.base64_encoder.encode(&sub_path, &random_bytes);
+
+    // find all preopens
+    const allocator = std.heap.page_allocator;
+    var wasi_preopens = try std.fs.wasi.preopensAlloc(allocator);
+
+    // look for the exact "/tmp" preopen match
+    const fd = std.fs.wasi.Preopens.find(wasi_preopens, path) orelse unreachable;
+    const base_dir = std.fs.Dir{ .fd = fd };
+
+    var tmp_path = base_dir.makeOpenPath(&sub_path, .{}) catch
+        @panic("unable to make tmp dir for testing: /tmp/<rand-path>");
+
+    defer tmp_path.close();
+    defer tmp_path.deleteTree(&sub_path) catch {};
+
+    // create a file under /tmp/<rand>/file.txt with contents "nonsense"
+    try tmp_path.writeFile(tmp_file_name, nonsense);
+
+    // now look for the file as a single path
+    var tmp_dir_path: [path.len + 1 + sub_path_len]u8 = undefined;
+
+    // find "/tmp/<rand>" using `findDir()`
+    _ = std.fmt.bufPrint(&tmp_dir_path, "{s}/{s}", .{ path, sub_path }) catch {};
```

Comment:
Same as above, rather than passing `tmp_dir_path` you should store the result of `bufPrint` and pass that. It will contain the correct length that was written to the buffer which means we don't have to pre-calculate those lengths above.

Reply:
same, this is a pattern I have seen in `lib/std/Build` :)